File: src/main/java/ca/rafemurray/FinishBlock.java
package ca.rafemurray;

import com.mojang.serialization.MapCodec;
import net.minecraft.block.BlockState;
import net.minecraft.block.BlockWithEntity;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.Entity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class FinishBlock extends BlockWithEntity {
    protected FinishBlock(Settings settings) {
        super(settings);
    }

    protected MapCodec<? extends BlockWithEntity> getCodec() {
        return createCodec(FinishBlock::new);
    }

    public BlockEntity createBlockEntity(BlockPos pos, BlockState state) {
        return new FinishBlockEntity(pos, state);
    }
    public void onSteppedOn(World world, BlockPos pos, BlockState state, Entity entity) {
        if ((world.getBlockEntity(pos) instanceof FinishBlockEntity finishBlockEntity)) {
            finishBlockEntity.onSteppedOn(entity);
        }
        super.onSteppedOn(world, pos, state, entity);
    }
}

File: src/main/java/ca/rafemurray/block/FinishBlock.kt
package ca.rafemurray.block

import ca.rafemurray.block.entity.FinishBlockEntity
import com.mojang.serialization.MapCodec
import net.minecraft.core.BlockPos
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.BaseEntityBlock
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState

class FinishBlock (settings: Properties) : BaseEntityBlock(settings) {
    override fun newBlockEntity(pos: BlockPos, state: BlockState): BlockEntity {
        return FinishBlockEntity(pos, state)
    }

    override fun stepOn(level: Level, pos: BlockPos, state: BlockState, entity: Entity) {
        val blockEntity = level.getBlockEntity(pos);
        if ((blockEntity is FinishBlockEntity)) {
            blockEntity.stepOn(entity)
        }
        super.stepOn(level, pos, state, entity)
    }

    override fun codec(): MapCodec<out BaseEntityBlock> {
        return simpleCodec { settings: Properties -> FinishBlock(settings) }
    }
}
